src/relations/db.py

```diff
@@ -0,0 +1,325 @@
+        """
+        if admin:
+            self.relation_name = "db-admin"
+        else:
+            self.relation_name = "db"
+
+        super().__init__(charm, self.relation_name)
+
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_departed, self._on_relation_departed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.admin = admin
+        self.charm = charm
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the legacy db/db-admin relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if not self.charm.unit.is_leader():
+            return
+
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+            or not self.charm.primary_endpoint
+        ):
+            event.defer()
+            return
+
+        logger.warning(f"DEPRECATION WARNING - `{self.relation_name}` is a legacy interface")
+
+        unit_relation_databag = event.relation.data[self.charm.unit]
+        application_relation_databag = event.relation.data[self.charm.app]
+
+        # Do not allow apps requesting extensions to be installed.
+        if "extensions" in unit_relation_databag or "extensions" in application_relation_databag:
+            logger.error(
+                "ERROR - `extensions` cannot be requested through relations"
+                " - they should be installed through a database charm config in the future"
+            )
+            self.charm.unit.status = BlockedStatus("extensions requested through relation")
+            return
+
+        # Sometimes a relation changed event is triggered,
+        # and it doesn't have a database name in it.
+        database = event.relation.data[event.app].get(
+            "database", event.relation.data[event.unit].get("database")
+        )
+        if not database:
+            logger.warning("No database name provided")
+            event.defer()
+            return
+
+        try:
+            # Creates the user and the database for this specific relation if it was not already
+            # created in a previous relation changed event.
+            user = f"relation-{event.relation.id}"
+            password = unit_relation_databag.get("password", new_password())
+            self.charm.postgresql.create_user(user, password, self.admin)
+            self.charm.postgresql.create_database(database, user)
+            postgresql_version = self.charm.postgresql.get_postgresql_version()
+        except (
```

Comment:
is it better to explicitly create a helper that checks if a user and database exists instead of using `try/except`?

Reply:
In this specific case, the errors wont happen if the user or database already exists. It is handled by the PostgreSQL library to check that and just [alter](https://github.com/canonical/postgresql-operator/blob/main/lib/charms/postgresql_k8s/v0/postgresql.py#L126) privileges in the user or database if they already exist. The errors will be raised if another unexpected thing happens.